lib/ConstExtract/ConstExtract.cpp

```diff
@@ -0,0 +1,239 @@
+//===-------- ConstExtract.pp -- Gather Compile-Time-Known Values --------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+#include "swift/ConstExtract/ConstExtract.h"
+#include "swift/AST/ASTContext.h"
+#include "swift/AST/ASTWalker.h"
+#include "swift/AST/Decl.h"
+#include "swift/AST/DiagnosticEngine.h"
+#include "swift/AST/DiagnosticsFrontend.h"
+#include "swift/AST/Evaluator.h"
+#include "swift/AST/SourceFile.h"
+#include "swift/AST/TypeCheckRequests.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/Support/JSON.h"
+
+#include <set>
+#include <sstream>
+#include <string>
+
+namespace {
+/// A helper class to collect all nominal type declarations that conform to
+/// AppIntents protocols.
+class NominalTypeConformanceCollector : public swift::ASTWalker {
+  const std::unordered_set<std::string> &Protocols;
+  std::vector<swift::NominalTypeDecl *> &ConformanceTypeDecls;
+
+public:
+  NominalTypeConformanceCollector(
+      const std::unordered_set<std::string> &Protocols,
+      std::vector<swift::NominalTypeDecl *> &ConformanceDecls)
+      : Protocols(Protocols), ConformanceTypeDecls(ConformanceDecls) {}
+
+  bool walkToDeclPre(swift::Decl *D) override {
+    if (auto *NTD = llvm::dyn_cast<swift::NominalTypeDecl>(D))
+      for (auto &Protocol : NTD->getAllProtocols())
```

Comment:
Just to double check my understanding, will `getAllProtocols()` also return the protocols that are conformed to in an extension?

Reply:
Great question. 
It appears that it will do just that. 
Modified the test in this PR to add the conformance separately in an extension, to capture this.